### PR TITLE
Set field length in metadata and fields warnings.

### PR DESCRIPTION
**Pull request by rgwozdz**
- Add ability to set field `length` property from provider metadata.
- Add warning if provider's field-metadata does not match feature properties.
- Refactored `field/index.js` to improve readability and implement code reuse.

### CODE REVIEW

File: test/utils.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
+const { detectType, esriTypeMap } = require('../src/utils')
+
+describe('Detect Types', () => {
+  describe('detectType', () => {
+    it('should return correct string representation of a values data type', () => {
+      detectType(1).should.equal('Integer')
+      detectType(100.1).should.equal('Double')
+      detectType('test').should.equal('String')
+      detectType(new Date()).should.equal('Date')
```
Comment (dmfenton) [inline test/utils.js]: we should also detect `new Date().toISOString()` as a date field

File: test/utils.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    })
+  })
+
+  describe('esriTypeMap', () => {
+    it('should return correct string representation of a values data type', () => {
+      esriTypeMap('Integer').should.equal('esriFieldTypeInteger')
+      esriTypeMap('Double').should.equal('esriFieldTypeDouble')
+      esriTypeMap('String').should.equal('esriFieldTypeString')
+      esriTypeMap('Date').should.equal('esriFieldTypeDate')
```
Comment (dmfenton) [inline test/utils.js]: there are some other types we should potentially add coverage for http://desktop.arcgis.com/en/arcobjects/latest/java/api/arcobjects/com/esri/arcgis/geodatabase/esriFieldType.html

File: src/query.js [hunk 3]
```diff
@@ -34,12 +34,18 @@ function query (data, params = {}) {
...
   if (options.toEsri) {
-    if (!hasIdField) console.warn(`The requested provider has no "idField" assignment. This can cause errors in ArcGIS clients`)
+    if (!hasIdField) console.warn(`WARNING: requested provider has no "idField" assignment. This can cause errors in ArcGIS clients`)
     else if (data.metadata.idField.toLowerCase() === 'objectid' && data.metadata.idField !== 'OBJECTID') {
-      console.warn(`The requested provider's "idField" is a mixed-case version of "OBJECTID". This can cause errors in ArcGIS clients`)
-    } else if (queriedData.features.some(feature => { return !Number.isInteger(feature.attributes.OBJECTID) || feature.attributes.isArrayOBJECTID > 2147483647 })) {
-      console.warn(`OBJECTIDs created from provider's "idField" are not integers from 0 to 2147483647`)
+      console.warn(`WARNING: requested provider's "idField" is a mixed-case version of "OBJECTID". This can cause errors in ArcGIS clients`)
     }
+    if (queriedData.features && queriedData.features.some(feature => { return !Number.isInteger(feature.attributes.OBJECTID) || feature.attributes.OBJECTID > 2147483647 })) {
```
Comment (dmfenton) [inline src/query.js]: this is a pretty expensive check, I think you'd be better off pushing this code into Winnow

File: src/query.js [hunk 4]
```diff
@@ -105,3 +111,32 @@ function queryStatistics (data, params) {
...
+  // compare metadata to feature properties; identifies fields defined in metadata that are not found in feature properties
+  // of that have a metadata type definition inconsistent with feature property's value
+  metadataFields.filter(field => {
+    let featureField = _.find(featureFields, ['name', field.name])
+    if (!featureField || (field.type !== featureField.type && !(field.type === 'Date' && featureField.type === 'Integer') && !(field.type === 'Double' && featureField.type === 'Integer'))) {
+      console.warn(`WARNING: requested provider's metadata field "${field.name} (${field.type})" not found in feature properties)`)
+    }
+  })
+
+  // compare feature properties to metadata fields; identfies fields found on feature that are not defined in metadata field array
```
Comment (dmfenton) [inline src/query.js]: typo `identfies`

File: src/query.js (hunk 3 above)
Comment (rgwozdz) [reply src/query.js]: Great catch, moved to winnow https://github.com/koopjs/winnow/pull/76.

File: test/utils.js (hunk 2 above)
Comment (rgwozdz) [reply test/utils.js]: Added in b15b98c613aa506bc19cbaa929d6d119a8c4f2d3.

File: src/utils.js [hunk 5]
```diff
@@ -72,6 +72,21 @@ function esriTypeMap (type) {
...
     case 'date':
       return 'esriFieldTypeDate'
+    case 'esrifieldtypeblob':
+    case 'esrifieldtypedate':
+    case 'esrifieldtypedouble':
+    case 'esrifieldtypegeometry':
+    case 'esrifieldtypeglobalid':
+    case 'esrifieldtypeguid':
+    case 'esrifieldtypeinteger':
+    case 'esrifieldtypeoid':
```
Comment (dmfenton) [inline src/utils.js]: I think we should follow that pattern we set where we remove `esriFieldType`

Comment (rgwozdz) [reply src/utils.js]: Is this a pattern currently found in FeatureServer? looking but don't see it

Comment (dmfenton) [reply src/utils.js]: it's in this very function.
Date => EsriFieldTypeDate
String => EsriFieldTypeString

Comment (rgwozdz) [reply src/utils.js]: Ok, misunderstood.  We should expect that the provider's metadata `fields` will omit the `esriFieldType` prefix, so we should omit it here as well.

Comment (dmfenton) [reply src/utils.js]: right. when we counsel people on how to use metadata, we should tell them to list fields without the `esriType` as this metadata might be used by other output types.

File: test/utils.js (hunk 2 above)
Comment (rgwozdz) [reply test/utils.js]: Update as per review comments in c0cc0f1a7d85247305f1c808479525dfc486bed5

**Pull request merged**